# Analyze NFL Play Type by team

<font color = red > The goal of the analysis will be to analyze the play calling of the Kansas City Chiefs football team. 

1. Download 2018 nfl play calling data from github
2. Review data to identify important data fields for analysis
3. Analyze pass play tendancy based on the down and distance of the KC Chiefs vs all NFL teams
4. Determine which NFL teams have the most pass-heavy play calling  


 </font>

## Step 1: download nfl play calling data from github

In [1]:
import pandas as pd # data manipulation library
import numpy as np # numerical computation library
import datetime as dt 

import matplotlib.pyplot as plt # plotting library
from matplotlib import cm # color maps for plotting
plt.style.use('ggplot') # use the ggplot plotting style


nfl = pd.read_csv('https://raw.githubusercontent.com/ryurko/nflscrapR-data/master/play_by_play_data/regular_season/reg_pbp_2017.csv', low_memory=False,error_bad_lines=False)





## Step 2: Review data to identify important data fields for analysis

In [2]:
bins = [0, 2, 5, 9.99,10, 15, np.inf]
names = ['<2', '2-5', '6-9', '10', '11-14','15+']

nfl['ydstogo_label'] = pd.cut(nfl['ydstogo'], bins=bins, labels=names)

chiefs = nfl.loc[nfl['posteam'] == "KC"]

In [3]:
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)

In [4]:
nfl.head()

play_id     game_id home_team away_team posteam posteam_type defteam  \
0       44  2017090700        NE        KC      NE         home      KC   
1       68  2017090700        NE        KC      NE         home      KC   
2       94  2017090700        NE        KC      NE         home      KC   
3      118  2017090700        NE        KC      NE         home      KC   
4      139  2017090700        NE        KC      NE         home      KC   

  side_of_field  yardline_100   game_date  quarter_seconds_remaining  \
0            KC          35.0  2017-09-07                      900.0   
1            NE          73.0  2017-09-07                      895.0   
2            NE          73.0  2017-09-07                      889.0   
3            NE          65.0  2017-09-07                      854.0   
4            NE          57.0  2017-09-07                      832.0   

   half_seconds_remaining  game_seconds_remaining game_half  quarter_end  \
0                  1800.0                  3600.0     Half1            0   
1                  1795.0                  3595.0     Half1            0   
2                  1789.0                  3589.0     Half1            0   
3                  1754.0                  3554.0     Half1            0   
4                  1732.0                  3532.0     Half1            0   

   drive  sp  qtr  down  goal_to_go   time  yrdln  ydstogo  ydsnet  \
0      1   0    1   NaN         0.0  15:00  KC 35        0      73   
1      1   0    1   1.0         0.0  14:55  NE 27       10       0   
2      1   0    1   2.0         0.0  14:49  NE 27       10       8   
3      1   0    1   3.0         0.0  14:14  NE 35        2      73   
4      1   0    1   1.0         0.0  13:52  NE 43       10      19   

                                                desc play_type  yards_gained  \
0  C.Santos kicks 64 yards from KC 35 to NE 1. D....   kickoff             0   
1  (14:55) NE 12-Brady 18th season as Patriots QB...      pass             0   
2  (14:49) T.Brady pass short right to R.Burkhead...      pass             8   
3  (14:14) (Shotgun) J.White left guard to NE 43 ...       run             8   
4  (13:52) (No Huddle, Shotgun) J.White up the mi...       run             3   

   shotgun  no_huddle  qb_dropback  qb_kneel  qb_spike  qb_scramble  \
0        0          0          0.0         0         0            0   
1        0          0          1.0         0         0            0   
2        0          0          1.0         0         0            0   
3        1          0          0.0         0         0            0   
4        1          1          0.0         0         0            0   

  pass_length pass_location  air_yards  yards_after_catch run_location  \
0         NaN           NaN        NaN                NaN          NaN   
1        deep          left       27.0                NaN          NaN   
2       short         right        1.0                7.0          NaN   
3         NaN           NaN        NaN                NaN         left   
4         NaN           NaN        NaN                NaN       middle   

  run_gap field_goal_result  kick_distance extra_point_result  \
0     NaN               NaN           64.0                NaN   
1     NaN               NaN            NaN                NaN   
2     NaN               NaN            NaN                NaN   
3   guard               NaN            NaN                NaN   
4     NaN               NaN            NaN                NaN   

  two_point_conv_result  home_timeouts_remaining  away_timeouts_remaining  \
0                   NaN                        3                        3   
1                   NaN                        3                        3   
2                   NaN                        3                        3   
3                   NaN                        3                        3   
4                   NaN                        3                        3   

   timeout timeout_team td_team  p

useful columns appear to be as follows: 

| Column | Description |
| --------------- | ----------- |
| ydstogo_label | e.g. <2, 2-5, 5-8, 8-9, 10 10-15, 15+|
| play_type	 | no_play, pass, kickoff, run|
| down | NaN, 0, 1, 2, 3,4|
| yardline_100	 | 1-99 with numbers > 50 on home side of field |
|posteam | team with possession|

## Step 3: Analyze pass play tendancy of the KC Chiefs vs all NFL teams

In [5]:
chiefs_runpassonly = chiefs[chiefs['play_type'].isin(['run','pass'])]

In [6]:
nfl_runpassonly = nfl[nfl['play_type'].isin(['run','pass'])]

In [7]:
pd.crosstab(chiefs_runpassonly.down,chiefs_runpassonly.play_type,margins = True, normalize='index')

play_type      pass       run
down                         
1.0        0.498871  0.501129
2.0        0.605590  0.394410
3.0        0.778894  0.221106
4.0        0.600000  0.400000
All        0.592402  0.407598

In [8]:
pd.crosstab(nfl_runpassonly.down,nfl_runpassonly.play_type,margins = True, normalize='index')

play_type      pass       run
down                         
1.0        0.476058  0.523942
2.0        0.577516  0.422484
3.0        0.799704  0.200296
4.0        0.646316  0.353684
All        0.581093  0.418907

> **Finding: Chiefs are more likely to call Pass Plays for all downs (59.2% vs 58.1%). Let's now look at play data based on down and distance**

In [9]:
chiefs_temp = pd.crosstab(chiefs_runpassonly.ydstogo_label,[chiefs_runpassonly.play_type,chiefs_runpassonly.down],normalize='index')

In [10]:
chiefs_temp2 = pd.crosstab(chiefs_runpassonly.ydstogo_label,chiefs_runpassonly.down,normalize='index')

In [11]:
chiefs_crosstab_play=chiefs_temp/chiefs_temp2

In [12]:
nfl_temp = pd.crosstab(nfl_runpassonly.ydstogo_label,[nfl_runpassonly.play_type,nfl_runpassonly.down],normalize='index')

In [13]:
nfl_temp2 = pd.crosstab(nfl_runpassonly.ydstogo_label,nfl_runpassonly.down,normalize='index')

In [14]:
nfl_crosstab_play=nfl_temp/nfl_temp2

In [15]:
chiefs_crosstab_play 

play_type          pass                                     run            \
down                1.0       2.0       3.0       4.0       1.0       2.0   
ydstogo_label                                                               
<2             0.333333  0.520000  0.275862  0.333333  0.666667  0.480000   
2-5            0.500000  0.500000  0.840000  0.666667  0.500000  0.500000   
6-9            0.444444  0.620690  0.873239  1.000000  0.555556  0.379310   
10             0.504950  0.644737  0.833333       NaN  0.495050  0.355263   
11-14          0.333333  0.742857  0.920000  0.000000  0.666667  0.257143   
15+            0.600000  0.500000  0.833333  1.000000  0.400000  0.500000   

play_type                          
down                3.0       4.0  
ydstogo_label                      
<2             0.724138  0.666667  
2-5            0.160000  0.333333  
6-9            0.126761  0.000000  
10             0.166667       NaN  
11-14          0.080000  1.000000  
15+            0.166667  0.000000

In [16]:
nfl_crosstab_play

play_type          pass                                     run            \
down                1.0       2.0       3.0       4.0       1.0       2.0   
ydstogo_label                                                               
<2             0.323671  0.328125  0.457261  0.347032  0.676329  0.671875   
2-5            0.407895  0.494867  0.843462  0.906250  0.592105  0.505133   
6-9            0.371795  0.595470  0.902377  0.923077  0.628205  0.404530   
10             0.477187  0.600937  0.920923  0.928571  0.522813  0.399063   
11-14          0.599099  0.722586  0.873684  0.857143  0.400901  0.277414   
15+            0.620000  0.780000  0.802846  0.842105  0.380000  0.220000   

play_type                          
down                3.0       4.0  
ydstogo_label                      
<2             0.542739  0.652968  
2-5            0.156538  0.093750  
6-9            0.097623  0.076923  
10             0.079077  0.071429  
11-14          0.126316  0.142857  
15+            0.197154  0.157895

In [17]:
chiefs_crosstab_play-nfl_crosstab_play

play_type          pass                                     run            \
down                1.0       2.0       3.0       4.0       1.0       2.0   
ydstogo_label                                                               
<2             0.009662  0.191875 -0.181399 -0.013699 -0.009662 -0.191875   
2-5            0.092105  0.005133 -0.003462 -0.239583 -0.092105 -0.005133   
6-9            0.072650  0.025219 -0.029138  0.076923 -0.072650 -0.025219   
10             0.027764  0.043800 -0.087589       NaN -0.027764 -0.043800   
11-14         -0.265766  0.020271  0.046316 -0.857143  0.265766 -0.020271   
15+           -0.020000 -0.280000  0.030488  0.157895  0.020000  0.280000   

play_type                          
down                3.0       4.0  
ydstogo_label                      
<2             0.181399  0.013699  
2-5            0.003462  0.239583  
6-9            0.029138 -0.076923  
10             0.087589       NaN  
11-14         -0.046316  0.857143  
15+           -0.030488 -0.157895

> **Summary of play pass tendancy on more common down and distance scenarios:**

| Down & Distance | Chiefs Pass % | NFL Pass % |
| --------------- | ----------- | ----------- |
| 1st & 10   | 55.5%    | 50.6% |
| 2nd & 10	 | 67.0% | 62.5% | 
| 2nd & 5    | 61.5% | 48.8% | 
| 2nd & 1	 | 41.7% | 34.7% | 
| 3rd & 10	 | 95.5% | 88.9% | 
| 3rd & 5    | 84.2% | 85.4% | 
| 3rd & 1	 | 51.3% | 43.2% | 

The chiefs call more pass plays than the average NFL team for all down/distance scenarios exccept 3rd & 5

2nd down & 5 yards is a scenario the Chiefs like to call pass plays and they do so at a much higher rate than the average NFL team (Chiefs 61.5% vs NFL team of 48.8%) 

Run plays are more likely in short yardage especially on 3rd down scenarios. However, even in 3rd and short the Chiefs are more likely to call pass plays (51%). 



## Step 4: Which NFL teams are the most pass-heavy

In [18]:
pd.crosstab(nfl_runpassonly.posteam,nfl_runpassonly.play_type,margins = True, normalize='index').sort_values(by='run')

play_type      pass       run
posteam                      
MIA        0.640530  0.359470
DET        0.635149  0.364851
TB         0.624514  0.375486
NYG        0.622935  0.377065
ARI        0.620921  0.379079
CLE        0.618382  0.381618
SF         0.617675  0.382325
OAK        0.617428  0.382572
GB         0.616090  0.383910
CIN        0.602876  0.397124
SEA        0.598577  0.401423
WAS        0.595876  0.404124
LAC        0.595382  0.404618
PIT        0.594752  0.405248
KC         0.592821  0.407179
NE         0.589354  0.410646
All        0.581440  0.418560
DEN        0.580433  0.419567
NYJ        0.572021  0.427979
BAL        0.569632  0.430368
PHI        0.566952  0.433048
ATL        0.565979  0.434021
NO         0.564995  0.435005
HOU        0.563601  0.436399
LA         0.556352  0.443648
TEN        0.551148  0.448852
IND        0.548980  0.451020
CHI        0.548596  0.451404
MIN        0.536680  0.463320
CAR        0.531281  0.468719
DAL        0.525100  0.474900
BUF        0.524194  0.475806
JAX        0.519700  0.480300

> **The Chiefs are the 10th most pass heavy team in the NFL**